web/auth/auth.go

```diff
@@ -177,145 +144,148 @@ func sessionData(session *sessions.Session) string {
+		req, err := http.NewRequest("GET", githubUserAPI, nil)
 		if err != nil {
-			logger.Error(err.Error())
-			return err
+			authenticationError(logger, w, fmt.Sprintf("failed to create user request: %s", err))
+			return
 		}
-		logger.Debugf("%s", sessionData(sess))
+		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", authToken.AccessToken))
 
-		// Try to get already logged in user.
-		// If session.Get(...) returns a new session, the check below will be nil.
-		if sess.Values[UserKey] != nil {
-			i, ok := sess.Values[UserKey]
-			if !ok {
-				logger.Debug("failed to get logged in user from session; logout")
-				return OAuth2Logout(logger)(c)
-			}
-			// If type assertions fails, the recover middleware will catch the panic and log a stack trace.
-			us := i.(*UserSession)
-			logger.Debug(us)
-			// Associate user with remote identity.
-			if err := db.AssociateUserWithRemoteIdentity(
-				us.ID, provider, remoteID, externalUser.AccessToken,
-			); err != nil {
-				logger.Debugf("Associate failed: %d, %s, %d, %s", us.ID, provider, remoteID, externalUser.AccessToken)
-				logger.Error("failed to associate user with remote identity", zap.Error(err))
-				return err
-			}
-			logger.Debugf("Associate: %d, %s, %d, %s", us.ID, provider, remoteID, externalUser.AccessToken)
+		resp, err := httpClient.Do(req)
+		if err != nil {
+			authenticationError(logger, w, fmt.Sprintf("failed to send user request: %v", err))
+			return
+		}
+		defer resp.Body.Close()
 
-			// Enable provider in session.
-			us.enableProvider(provider)
-			if err := sess.Save(r, w); err != nil {
-				logger.Error(err.Error())
-				return err
-			}
-			logger.Debugf("enableProvider: %v, r=%v, w=%v", provider, r, w)
-
-			if user, err := db.GetUser(us.ID); err != nil {
-				// handle
-				logger.Errorf("Failed to get user: %v", err)
-			} else {
-				if ok := updateScm(c, logger, scms, user); !ok {
-					logger.Debugf("Failed to update SCM for User: %v", user)
-				}
-			}
+		if resp.StatusCode != http.StatusOK {
+			authenticationError(logger, w, fmt.Sprintf("unexpected OAuth provider response: %d (%s)", resp.StatusCode, resp.Status))
+			return
+		}
+		responseBody, err := io.ReadAll(resp.Body)
+		if err != nil {
+			authenticationError(logger, w, fmt.Sprintf("failed to read uthentication response: %v", err))
+			return
+		}
 
-			// Enable gRPC requests for session
-			if token := extractSessionCookie(w); len(token) > 0 {
-				logger.Debugf("extractSessionCookie: %v", token)
-				Add(token, us.ID)
-			} else {
-				logger.Debugf("no session cookie found in w: %v", w)
-			}
-			logger.Debugf("Redirecting: %s", redirect)
-			return c.Redirect(http.StatusFound, redirect)
+		// Complete user authentication.
+		externalUser := &externalUser{}
+		if err := json.NewDecoder(bytes.NewReader(responseBody)).Decode(&externalUser); err != nil {
+			authenticationError(logger, w, fmt.Sprintf("failed to decode user information: %v", err))
+			return
 		}
```

Comment:
Code from lines 244-274 could be made into a function `fetchExternalUser() (*externalUser, error)` or `fetchRemoteID()` if you include lines 287. But then you maybe don't want to log the JSON output... (since I wouldn't suggest to pass in the logger to this method...)